src/tchannel_bridge.js

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import * as constants from './constants';
+import Span from './span';
+import SpanContext from './span_context';
+import Utils from './util';
+import opentracing from 'opentracing';
+import Tracer from '../src/tracer';
+import TextMapCodec from '../src/propagators/text_map_codec';
+
+let TCHANNEL_TRACING_PREFIX = '$tracing$';
+
+export default class TChannelBridge {
+    _tracer: Tracer;
+    _codec: TextMapCodec;
+
+    constructor(tracer: Tracer) {
+        this._tracer = tracer;
+        this._codec = new TextMapCodec({
+            urlEncoding: false,
+            contextKey: TCHANNEL_TRACING_PREFIX + constants.TRACER_STATE_HEADER_NAME,
+            baggagePrefix: TCHANNEL_TRACING_PREFIX + constants.TRACER_BAGGAGE_HEADER_PREFIX
+        });
+    }
+
+    _tchannelCallbackWrapper(span, wrappedCallback, err, res) {
+        if (err) {
+            span.setTag(opentracing.Tags.ERROR, true);
+        }
+
+        span.finish();
+        return wrappedCallback(err, res);
+    }
+
+    /**
+     * Wraps a tchannel handler, and takes a context in order to populate the incoming context
+     * with a span.
+     *
+     * @param {Function} [handlerFunc] - a tchannel handler function that responds to an incoming request.
+     * @param {Object} [options] - options to be passed to a span on creation.
+     * @returns {Function} - a function that wrapps the handler in order to automatically populate
+     * a the handler's context with a span.
+     **/
+    tracedHandler(handlerFunc: any, options: startSpanArgs = {}): Function {
+        return (context, request, headers, body, wrappedCallback) => {
+            let operationName = options.operationName || request.arg1;
+            let span = this._extractSpan(operationName, headers);
+            span.setTag(opentracing.Tags.SPAN_KIND, opentracing.Tags.SPAN_KIND_RPC_SERVER);
```

Comment:
OpenTracing conventions is to pass RPC_SERVER kind into the startSpan method, not set it afterwards